template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmHostGroupRoleConfigProviders.java

```diff
@@ -0,0 +1,158 @@
+
+import com.cloudera.api.swagger.model.ApiClusterTemplateConfig;
+import com.cloudera.api.swagger.model.ApiClusterTemplateHostTemplate;
+import com.cloudera.api.swagger.model.ApiClusterTemplateInstantiator;
+import com.cloudera.api.swagger.model.ApiClusterTemplateRoleConfigGroup;
+import com.cloudera.api.swagger.model.ApiClusterTemplateRoleConfigGroupInfo;
+import com.cloudera.api.swagger.model.ApiClusterTemplateService;
+import com.sequenceiq.cloudbreak.template.TemplatePreparationObject;
+import com.sequenceiq.cloudbreak.template.model.ServiceComponent;
+import com.sequenceiq.cloudbreak.template.views.HostgroupView;
+
+@Component
+public class CmHostGroupRoleConfigProviders {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CmHostGroupRoleConfigProviders.class);
+
+    @Inject
+    private List<CmHostGroupRoleConfigProvider> providers;
+
+    public void process(CmTemplateProcessor templateProcessor, TemplatePreparationObject source) {
+        if (!getHostTemplates(templateProcessor).isEmpty()) {
+            updateConfigsInTemplate(templateProcessor, generateConfigs(templateProcessor, source));
+        }
+    }
+
+    private List<ApiClusterTemplateHostTemplate> getHostTemplates(CmTemplateProcessor templateProcessor) {
+        return ofNullable(templateProcessor.getTemplate().getHostTemplates()).orElseGet(List::of);
+    }
+
+    private Map<String, Map<String, List<ApiClusterTemplateConfig>>> generateConfigs(CmTemplateProcessor templateProcessor, TemplatePreparationObject source) {
+        Map<String, Map<String, List<ApiClusterTemplateConfig>>> configsByRoleConfigGroup = new HashMap<>();
+
+        Map<String, HostgroupView> hostGroups = source.getHostgroupViews().stream()
+                .collect(toMap(HostgroupView::getName, Function.identity()));
+        List<ApiClusterTemplateHostTemplate> hostTemplates = getHostTemplates(templateProcessor);
+        Map<String, ServiceComponent> serviceComponents = templateProcessor.mapRoleRefsToServiceComponents();
+
+        for (ApiClusterTemplateHostTemplate hostTemplate : hostTemplates) {
+            String hostGroupName = hostTemplate.getRefName();
+            List<String> roleConfigGroups = ofNullable(hostTemplate.getRoleConfigGroupsRefNames()).orElseGet(List::of);
+            HostgroupView hostgroupView = hostGroups.get(hostGroupName);
+
+            for (String roleConfigGroup : roleConfigGroups) {
+                for (CmHostGroupRoleConfigProvider provider : providers) {
+                    ServiceComponent serviceComponent = serviceComponents.get(roleConfigGroup);
+
+                    if (serviceComponent != null
+                            && Objects.equals(provider.getServiceType(), serviceComponent.getService())
+                            && provider.getRoleTypes().contains(serviceComponent.getComponent())) {
+
+                        configsByRoleConfigGroup.computeIfAbsent(roleConfigGroup, __ -> new HashMap<>())
```

Comment:
Yes, because [`computeIfAbsent` expects a `Function`](https://docs.oracle.com/javase/8/docs/api/java/util/HashMap.html#computeIfAbsent-K-java.util.function.Function-), while `HashMap::new` is a `Supplier` (no input).

Reply:
OK.